File: client/src/test/java/com/jraska/github/client/ui/RepoDetailActivityTest.java
package com.jraska.github.client.ui;

import android.app.Activity;
import android.arch.lifecycle.ViewModel;
import android.content.Intent;
import com.jraska.github.client.BuildConfig;
import com.jraska.github.client.Navigator;
import com.jraska.github.client.R;
import com.jraska.github.client.ViewModelFactoryDecorator;
import com.jraska.github.client.analytics.EventAnalytics;
import com.jraska.github.client.rx.AppSchedulers;
import com.jraska.github.client.users.RepoDetailViewModel;
import com.jraska.github.client.users.UserViewModelModule;
import com.jraska.github.client.users.UsersRepository;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import okhttp3.HttpUrl;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class RepoDetailActivityTest {

  @Captor ArgumentCaptor<Intent> intentCaptor;
  @Mock Activity activityMock;
  @Mock UsersRepository repositoryMock;
  @Mock Navigator navigatorMock;

  @Before
  public void setUp() {
    MockitoAnnotations.initMocks(this);

    when(repositoryMock.getRepoDetail(any(), any())).thenReturn(Observable.empty());

    ViewModel viewModel = UserViewModelModule.provideRepoDetailModel(repositoryMock, trampoline(),
      navigatorMock, EventAnalytics.EMPTY);

    ViewModelFactoryDecorator.setToApp(RepoDetailViewModel.class, viewModel);
  }

  @Test
  public void whenFabClicked_thenNavigatesToGitHub() {
    RepoDetailActivity.start(activityMock, "jraska/Falcon");
    verify(activityMock).startActivity(intentCaptor.capture());

    RepoDetailActivity activity = Robolectric.buildActivity(RepoDetailActivity.class,
      intentCaptor.getValue()).setup().get();

    activity.findViewById(R.id.repo_detail_github_fab).performClick();

    verify(navigatorMock).launchOnWeb(HttpUrl.parse("https://github.com/jraska/Falconn"));
  }

  public static AppSchedulers trampoline() {
    return new AppSchedulers(Schedulers.trampoline(), Schedulers.trampoline(),
      Schedulers.trampoline());
  }
}

File: client/src/androidTest/java/com/jraska/github/client/users/RepoDetailActivityTest.kt
package com.jraska.github.client.users

import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.ViewActions
import android.support.test.espresso.matcher.ViewMatchers.withId
import com.jraska.github.client.*
import com.jraska.github.client.analytics.EventAnalytics
import io.reactivex.Observable
import okhttp3.HttpUrl
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations

class RepoDetailActivityTest {
  @Mock lateinit var navigatorMock: Navigator
  @Mock lateinit var repositoryMock: UsersRepository

  @get:Rule
  var passRule = MakeTestsPassRule()

  @Before
  fun before() {
    MockitoAnnotations.initMocks(this)

    `when`(repositoryMock.getRepoDetail("jraska", "Falcon")).thenReturn(Observable.empty())
  }

  @Test
  fun whenFabClicked_thenNavigatesToGitHub() {
    val detailViewModel = RepoDetailViewModel(repositoryMock,
      AppModule.schedulers(), navigatorMock, EventAnalytics.EMPTY)
    ViewModelFactoryDecorator.setToApp(RepoDetailViewModel::class.java, detailViewModel)

    val deepLink = "https://github.com/jraska/Falcon"
    DeepLinkLaunchTest.launchDeepLink(deepLink)

    onView(withId(R.id.repo_detail_github_fab)).perform(ViewActions.click())
    verify<Navigator>(navigatorMock).launchOnWeb(HttpUrl.parse(deepLink)!!)
  }
}
